### PR TITLE
Separate cli arg parsing from initialization

### LINKED ISSUE #1167: ClimateMachine.init() should not parse process ARGS by default?
# Description

Currently ClimateMachine parses command line arguments by default upon initialization: https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Driver/Driver.jl#L212.

This has the side effect of exclusively claiming process global `ARGS`, which doesn't compose well when invoked in a programmatic / sub-procedure context (ex. it makes creating an alternative CLI program that invokes ClimateMachine awkward to write).  One example of this behaviour is that it breaks running a subset of tests with the current test harness: 

`julia --project=@. -e 'using Pkg; Pkg.test(test_args=["Numerics"])`

```
Starting tests for Numerics
Starting tests for Mesh
┌ Warning: Incompatibility detected between CUDA and LLVM 8.0+; disabling debug info emission for CUDA kernels
└ @ CUDAnative ~/.julia/packages/CUDAnative/hfulr/src/CUDAnative.jl:114
too many arguments
usage: <PROGRAM> [--disable-gpu] [--show-updates <interval>]
                 [--diagnostics <interval>] [--vtk <interval>]
                 [--monitor-timestep-duration <interval>]
                 [--monitor-courant-numbers <interval>]
                 [--checkpoint <interval>] [--checkpoint-keep-all]
                 [--checkpoint-at-end] [--checkpoint-dir <path>]
                 [--restart-from-num <number>] [--fix-rng-seed]
                 [--log-level <level>] [--output-dir <path>]
                 [--integration-testing]
ERROR: Package ClimateMachine errored during testing
```

### Proposal

`ClimateMachine.int()` should be configured through keyword arguments and the public API should not depend on passing ArgParse data structures. Methods with overloads for simple cases like passing a dictionary of values could be added (ex. json / TOML processing).  `ClimateMachine.cli()` would exhibit the current behaviour of parsing process `ARGS` and then internally invoking `ClimateMachine.init()`.

# For CLIMA Developers

- [ ] This feature can be added (if it cannot be, explain why in a comment below -- lack of technical expertise, not relevant to the scope of this project, too ambitious)
- [ ] There is a timeline for when this feature can be implemented
- [ ] The feature has been (or will be) implemented (Please link the PR)

### PR DESCRIPTION
**Pull request by jakebolewski**
# Description

Separates out cli argument processing from `ClimateMachine.init()` by defining a new method `ClimateMachine.cli()` that parses arguments and initializes ClimateMachine runtime.  `ClimateMachine.init()` is preserved to allow for programmatic initialization when command line parsing is not needed (tests / notebooks / scripts).  `ClimateMachine.cli` and `ClimateMachine.init` now take keyword arguments to allow for easy overloading for `ClimateMachine.Settings` at configuration / runtime.  The experiment files have been updated to use `cli`, the tutorials I kept the same. 

Consistent overloading fallback behaviour is also defined: 

1. cli arguments (if applicable)
2.  keyword argument overloads
3. `CLIMATEMACHINE_SETTINGS_<VALUE_NAME>` process environment definitions
4. `ClimateMachine.Settings` default values. 

This allows for easy overloading of settings in batch scripts / container environments without having to define configuration files or change the driver file source code.  We were currently doing this in kind of an add-hoc fashion.

Closes #1167 

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://CliMA.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://CliMA.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Driver/Driver.jl [hunk 1]
```diff
@@ -107,32 +137,64 @@ function parse_commandline(custom_settings)
         help = "interval at which to show simulation updates"
         metavar = "<interval>"
         arg_type = String
-        default = "60secs"
+        default = if haskey(defaults, :show_updates)
+            defaults[:show_updates]
+        else
+            get(ENV, "CLIMATEMACHINE_SETTINGS_SHOW_UPDATES", Settings.show_updates)
+        end
```
Comment (kpamnany) [inline src/Driver/Driver.jl:144]: Maybe something like the following? A function might be more compact...
```
macro get_default(setting, defaults_dict, defaults_struct)
    quote
        let setting = $(esc(setting)),
            defaults_dict = $(esc(defaults_dict)),
            defaults_struct = $(esc(defaults_struct)),
            setting_sym = Symbol(setting),
            setting_upper = uppercase(setting),
            env_var = string("CLIMATEMACHINE_SETTINGS_", setting_upper)
        if haskey(defaults_dict, setting_sym)
            defaults_dict[setting_sym]
        else
            get(ENV, env_var, Base.getproperty(defaults_struct, setting_sym))
        end
        end
    end
end
```

File: src/Driver/Driver.jl [hunk 2]
```diff
@@ -143,53 +205,108 @@ function parse_commandline(custom_settings)
...
-    ClimateMachine.array_type()
+    ClimateMachine.cli(;arg_settings=nothing, init_driver=true, kwargs...)
 
-Return the array type used by ClimateMachine. This defaults to (CPU-based) `Array`
-and is only correctly set (based on choice from the command line, from
-an environment variable, or from experiment code) after `ClimateMachine.init()`
-is called.
+Initialize the ClimateMachine runtime with cli argument parsing.
+
+- Additional `ArgParseSetting` behavior can be injected into the default
```
Comment (kpamnany) [inline src/Driver/Driver.jl:274]:
```suggestion
- Additional `ArgParseSettings` behavior can be injected into the default
```

File: src/Driver/Driver.jl [hunk 3]
```diff
@@ -143,53 +205,108 @@ function parse_commandline(custom_settings)
...
+ClimateMachine `ArgParseSetting` cofiguration by setting the `arg_settings`
```
Comment (kpamnany) [inline src/Driver/Driver.jl:275]:
```suggestion
ClimateMachine `ArgParseSettings` cofiguration by setting the `custom_settings`
```

File: src/Driver/Driver.jl [hunk 4]
```diff
@@ -143,53 +205,108 @@ function parse_commandline(custom_settings)
...
+keyword value.
+
+- Setting `init_driver=false` will set the ClimateMachine.Setting singleton
```
Comment (kpamnany) [inline src/Driver/Driver.jl:278]:
```suggestion
- Setting `init_driver = false` will set the ClimateMachine.Settings singleton
```

File: src/Driver/Driver.jl [hunk 5]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
-function init(; disable_gpu = false, arg_settings = nothing)
-    # set up timing mechanism
-    tictoc()
+function init(; init_driver::Bool = true, kwargs...)
+    # init global setting values
+    # TODO: add validation for initialization values
+    if haskey(kwargs, :disable_gpu)
+        Settings.disable_gpu = kwargs[:disable_gpu]
+    elseif haskey(ENV, "CLIMATEMACHINE_GPU")
+        #TODO: included for backwards compat, move to shared prefix scheme
```
Comment (kpamnany) [inline src/Driver/Driver.jl:366]:
```suggestion
        @warn("CLIMATEMACHINE_GPU will be deprecated; use CLIMATEMACHINE_SETTINGS_DISABLE_GPU")
```

File: src/Driver/Driver.jl (hunk 4 above)
Comment (kpamnany) [reply src/Driver/Driver.jl:278]: Also some similar below.

File: src/Driver/Driver.jl [hunk 6]
```diff
@@ -143,53 +205,108 @@ function parse_commandline(custom_settings)
...
+function cli(;
+    custom_settings::Union{Nothing, ArgParseSettings} = nothing,
+    init_driver::Bool = true,
+    kwargs...,
+)
+    kw_defaults = Dict{Symbol, Any}(kwargs)
+    parsed_args = parse_commandline(kw_defaults, custom_settings)
+    # we need to munge the parsed arg dict a bit as parsed arg keys
+    # and climatemachine initialization keywords are not 1:1
+    parsed_args["checkpoint_keep_one"] = !parsed_args["checkpoint_keep_all"]
```
Comment (kpamnany) [inline src/Driver/Driver.jl:298]: No idea why I did this. Let's change it so there isn't such an exception? In this file and in `Utilities/Callbacks/Callbacks.jl`.

File: src/Driver/Driver.jl [hunk 7]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
-        Settings.vtk = parsed_args["vtk"]
-        delete!(parsed_args, "vtk")
-        Settings.monitor_timestep_duration =
-            parsed_args["monitor-timestep-duration"]
-        delete!(parsed_args, "monitor-timestep-duration")
+    if haskey(kwargs, :vtk)
+        Settings.vtk = kwargs[:vtk]
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_VTK")
+        Settings.vtk = get(ENV, "CLIMATEMACHINE_SETTINGS_VTK")
+    end
```
Comment (kpamnany) [inline src/Driver/Driver.jl:389]: Another macro, as above, or a function?

File: src/Driver/Driver.jl [hunk 8]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
+    if haskey(kwargs, :restart_from_num)
+        Settings.restart_from_num = kwargs[:restart_from_num]
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_RESTART_FROM_NUM")
+        Settings.restart_from_num =
+            parse(Int, get(ENV, "CLIMATEMACHINE_SETTINGS_RESTART_FROM_NUM"))
     end
 
+    if haskey(kwargs, :fix_rng_seed)
+        Settings.fix_rng_seed = kwargs[:fix_rng_seed]
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_FIX_RNG")
```
Comment (kpamnany) [inline src/Driver/Driver.jl:442]:
```suggestion
    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_FIX_RNG_SEED")
```

File: src/Driver/Driver.jl [hunk 9]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
+        Settings.fix_rng_seed =
+            parse(Bool, get(ENV, "CLIMATEMACHINE_SETTINGS_FIX_RNG"))
```
Comment (kpamnany) [inline src/Driver/Driver.jl:444]:
```suggestion
            parse(Bool, get(ENV, "CLIMATEMACHINE_SETTINGS_FIX_RNG_SEED"))
```

File: src/Driver/Driver.jl [hunk 10]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
+        Settings.log_level = uppercase(kwargs[:log_level])
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_LOG_LEVEL")
+        Settings.log_level =
+            uppercase(get(ENV, "CLIMATEMACHINE_SETTINGS_LOG_LEVEL"))
+    end
+
+    if haskey(kwargs, :output_dir)
+        Settings.output_dir = kwargs[:output_dir]
+    elseif haskey(ENV, "CLIMATEMACHINE_OUTPUT_DIR")
+        Settins.output_dir = get(ENV, "CLIMATEMACHINE_OUTPUT_DIR")
```
Comment (kpamnany) [inline src/Driver/Driver.jl:457]:
```suggestion
    elseif haskey(ENV, "CLIMATEMACHINE_OUTPUT_DIR")
        @warn("CLIMATEMACHINE_OUTPUT_DIR will be deprecated; use CLIMATEMACHINE_SETTINGS_OUTPUT_DIR")
        Settings.output_dir = get(ENV, "CLIMATEMACHINE_OUTPUT_DIR")
```

File: src/Driver/Driver.jl [hunk 11]
```diff
@@ -200,68 +317,181 @@ it will be imported into ClimateMachine's settings.
...
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_OUTPUT_DIR")
+        Settings.output_dir = get(ENV, "CLIMATEMACHINE_SETTINGS_OUTPUT_DIR")
+    end
+
+    if haskey(kwargs, :integration_testing)
+        Settings.integration_testing = kwargs[:integration_testing]
+    elseif haskey(ENV, "CLIMATEMACHINE_SETTINGS_INTEGRATION_TESTING")
+        Settings.integration_testing =
+            get(ENV, "CLIMATEMACHINE_INTEGRATION_TESTING")
```
Comment (kpamnany) [inline src/Driver/Driver.jl:466]:
```suggestion
            get(ENV, "CLIMATEMACHINE_SETTINGS_INTEGRATION_TESTING")
```

**Comment by kpamnany** [comment]
Worth checking SlurmCI results, since a couple of different command line arguments are used.

bors try

File: src/Driver/Driver.jl [hunk 12]
```diff
@@ -67,20 +68,48 @@ Base.@kwdef mutable struct ClimateMachine_Settings
...
+If `defaults` is not nothing, override default values for cli argument defaults.
+If `custom_settings` arg is not nothing, add ArgParseSettings to the parsing step.
+
+Returns a `Dict` containing parsed process ARGS values.
+"""
+function parse_commandline(
+    defaults::Union{Nothing, Dict{Symbol, Any}} = nothing,
+    custom_settings::Union{Nothing, ArgParseSettings} = nothing,
+)
+    if defaults == nothing
```
[the remark this answers is not in the document]
Comment (jakebolewski) [reply src/Driver/Driver.jl]: good point, `val == nothing` is just less efficient

File: src/Driver/Driver.jl (hunk 1 above)
Comment (jakebolewski) [reply src/Driver/Driver.jl:144]: In general I loathe macros, I feel they are overused :) A function would be better. There is some custom processing / validation that need to happen for each setting so I don't know if it's worth it to consolidate, so the loc gain would be minimal at the moment.  maybe something to do later when there are more settings depending on how this evolves?

**Comment by jakebolewski** [comment]
I went with `CLIMATEMACHINE_SETTINGS_` to make it easier to grep / aggregate the logs.  Invariably there will probably be more ENV variables that change different behaviours so probably there should be an easy way to group them.

**Comment by jakebolewski** [comment]
bors try

**Comment by blallen** [comment]
Will this let us use command-line arguments when running CI tests on azure and/or slurm? If so, that will be a great boon to simplifying the number of different test files we have.

File: src/Driver/Driver.jl (hunk 1 above)
Comment (kpamnany) [reply src/Driver/Driver.jl:144]: Ah, I see the `parse(Int, ...)` and the `uppercase()...`, yeah. Messy, but no way out. :-/

**Comment by blallen** [comment]
> @blallen do you have some examples? I think maybe (the need for this was born out of decoupling some of the test runners from overloading argument parsing).

For the ocean model, `experiments/OceanBoxGCM/ocean_gyre.jl` and `test/Ocean/HydrostaticBoussinesq/test_ocean_gyre.jl` are basically the same file, but with different numbers of elements, integration times, and sets of boundary conditions. If this works on both slurm and azure, we can only keep the `experiments` version and change the differences between the two using command-line arguments. It will also make the experiment file more user friendly, as currently, you have to modify the file to actually run just a single experiment offline instead of running 8 variations as we do for the slurm CI test.

**Comment by kpamnany** [comment]
I think this is what the `--integration-testing` argument is for Brandon...?

**Comment by jakebolewski** [comment]
bors try

File: src/Driver/Driver.jl (hunk 1 above)
Comment (charleskawczynski) [reply src/Driver/Driver.jl:144]: What about

```julia
function get_default(Settings, defaults, sym)
    default = if haskey(defaults, sym)
        defaults[sym]
    else
        get(
            ENV,
            "CLIMATEMACHINE_SETTINGS_"*upper(String(sym)),
            getproperty(Settings, sym),
        )
    end
end
```

Comment (jakebolewski) [reply src/Driver/Driver.jl:144]: I think we will need to move to a parameter framework that just generalizes everything so you can support things like running with a config / env overloads/ cli generation / and run time configuration that would encapsulate a lot of these details.  Maybe something like this already exists in the Julia world that we can re-use.  I don't want to do that right now though, or if it's worth it until other details get sorted out.

Comment (charleskawczynski) [reply src/Driver/Driver.jl:144]: I guess the question is, how many custom cases do we need to account for? If it's fewer than 5, I think it's worth writing some helper scripts / using existing Julia tools to simplify things. But I'm fine tackling this in a later PR.

Comment (kpamnany) [reply src/Driver/Driver.jl:144]: > I think we will need to move to a parameter framework that just generalizes everything so you can support things like running with a config / env overloads/ cli generation / and run time configuration that would encapsulate a lot of these details. Maybe something like this already exists in the Julia world that we can re-use. I don't want to do that right now though, or if it's worth it until other details get sorted out.

Absolutely. 💯  See #947 for where we need to take this entire layer at some point.

**Comment by jakebolewski** [comment]
@charleskawczynski I consolidated a bit with about a 50 loc reduction so not nothing, if we get rid of the special cases then we probably can just use reflection while things are simple / easy.  

I think eventually the special cases and need for better errors will probably dictate something more sophisticated.

**Comment by blallen** [comment]
> I think this is what the `--integration-testing` argument is for Brandon...?

Ah, didn't know this existed. This will cover making the default behavior for the users nice, but not running multiple tests out of one script. But I should probably make a PR using the `--integration-testing`

**Comment by kpamnany** [comment]
> Ah, didn't know this existed. This will cover making the default behavior for the users nice, but not running multiple tests out of one script. But I should probably make a PR using the `--integration-testing`

A number of the `Numerics/` tests are using it, if you'd like to see examples.

As for the multiple tests from one script, you can use the Driver code as a template for making an `ArgParse.ArgParseSettings`, pass it into `ClimateMachine.cli()` and get your CLI arguments back.

Or you can parse `ARGS` in whatever way you choose in your experiment file and use `ClimateMachine.init() to send the general `ClimateMachine` arguments into the Driver, but this will break what Jake's trying to do here maybe?

File: src/Driver/Driver.jl [hunk 13]
```diff
@@ -143,53 +210,84 @@ function parse_commandline(custom_settings)
...
+- ClimateMachine.init key value pairs can be supplied to overload
+default system defaults at runtime, default values will be merged with the
+parsed argument settings, with parsed cli argument values taking precedent
+over runtime defined default values.
+
+
+Returns a `Dict` containing parsed process ARGS values.
 """
-array_type() = Settings.array_type
+function cli(;
```
Comment (simonbyrne) [inline src/Driver/Driver.jl:270]: should we call this `init_cli`? or `init_args`?

**Comment by jakebolewski** [comment]
bors try

File: src/Driver/Driver.jl [hunk 14]
```diff
@@ -143,53 +210,84 @@ function parse_commandline(custom_settings)
...
         metavar = "<path>"
         arg_type = String
-        default = get(ENV, "CLIMATEMACHINE_OUTPUT_DIR", "output")
+        default = if haskey(defaults, :output_dir)
+            defaults[:output_dir]
+        elseif haskey(ENV, "CLIMATEMACHINE_OUTPUT_DIR")
+            ENV["CLIMATEACHINE_OUTPUT_DIR"]
+        else
+            get(ENV, "CLIMATEMACHINE_SETTINGS_OUTPUT_DIR", Settings.output_dir)
+        end
```
Comment (simonbyrne) [inline src/Driver/Driver.jl:237]:
```suggestion
        default = get(defaults, :output_dir) do
            get(ENV, "CLIMATEMACHINE_OUTPUT_DIR") do
                get(ENV, "CLIMATEMACHINE_SETTINGS_OUTPUT_DIR", Settings.output_dir)
            end
        end
```

**Comment by jakebolewski** [comment]
bors try

**Comment by jakebolewski** [comment]
bors try

File: src/Driver/Driver.jl [hunk 15]
```diff
@@ -67,20 +68,79 @@ Base.@kwdef mutable struct ClimateMachine_Settings
...
+    if haskey(settings, setting_name)
+        return convert(setting_type, settings[setting_name])
+    elseif haskey(ENV, setting_env)
+        env_val = ENV[setting_env]
+        v = tryparse(setting_type, env_val)
+        if isnothing(v)
+            error("Cannot parse ENV $setting_env value $env_val, to setting type $setting_type")
+        end
+        return v
+    elseif haskey(defaults, setting_name)
```
Comment (kpamnany) [inline src/Driver/Driver.jl:116]:
```suggestion
    else
```

File: src/Driver/Driver.jl [hunk 16]
```diff
@@ -67,20 +68,79 @@ Base.@kwdef mutable struct ClimateMachine_Settings
...
+        return defaults[setting_name]
+    else
+        error("setting $setting_name is not contained in either settings or defaults")
+    end
```
Comment (kpamnany) [inline src/Driver/Driver.jl:120]:
```suggestion
    end
```
